src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs

```diff
@@ -915,7 +920,10 @@ protected override void ProcessRecord()
                             Path = SessionState.Internal.GetSingleProvider(Commands.FileSystemProvider.ProviderName).Home;
                         }
 
+                        Path = ReplacePathWithLastLocationIfApplicable(Path);
+                        var initialPath = SessionState.Path.CurrentLocation.Path;
                         result = SessionState.Path.SetLocation(Path, CmdletProviderContext);
+                        Environment.SetEnvironmentVariable(environmentVariableNameForPreviousLocation, initialPath);
```

Comment:
We shouldn't use environment variables, instead we should expose a new variable (`PSOldPWD`?) that is an instance of a bounded stack.

We could consider a user configurable stack size, but we can leave that for later (not this PR).

I also wonder if this logic should be in `SessionState.Path.SetLocation` - that could be useful for alternate implementations of `Set-Location` that call the api.